src/libexpr/eval-inline.hh

```diff
@@ -52,26 +52,26 @@ void EvalState::forceValue(Value & v, Callable getPos)
 }
 
 
-inline void EvalState::forceAttrs(Value & v, const Pos & pos)
+inline void EvalState::forceAttrs(Value & v, const Pos & pos, const std::string & errorCtx)
```

Comment:
this (and presumably most string params you've added) should be `const char *`s instead where at all possible. that saves a strlen and copy of the string to dynamic memory on each call, which is presumably makes for the majority of the performance impact.

Reply:
`std::string_view` should be fine too?

Reply:
`string_view` also calls `strlen` when constructed from a string literal, so that's also mostly unnecessary overhead.

Reply:
https://en.cppreference.com/w/cpp/string/basic_string_view/basic_string_view I don't think that's true?

Reply:
constructing from a string literal will use overload 4 in the current code base, which calls a strlen-equivalent via the traits type. a strlen at runtime can only be avoided if the length operation of the traits type is itself constexpr, which doesn't seem to be part of the spec. if it is constexpr it's totally possible that you're right and the strlen is elided. :)

Reply:
Whew! I do agree the number of things that need to go right for this to work as intended is very frightening.

I guess always constexpr bind any constant string expression to be safe rather than plopping the expression right in the function argument slot to be safe.

Reply:
as long as it doesn't eat our hard-won performance gains we won't complain :) would be nice to have an interface that guarantees things not wasting resources though. `const char *` would do that, or maybe a new type that only accepts `const char (&)[N]` on construction?

Reply:
There is `std::array` but I am not sure how to get `std::string_view` from it in a nice way.
`constexpr std::string_view msg = "asdfasdfds";` I think is good enough.